packages/terra-functional-testing/lib/services/wdio-docker-service.js

```diff
@@ -0,0 +1,167 @@
+/* eslint-disable class-methods-use-this */
+const path = require('path');
+const util = require('util');
+const childProcess = require('child_process');
+const Logger = require('../logger/logger');
+
+const exec = util.promisify(childProcess.exec);
+const logger = new Logger({ prefix: 'wdio-docker-service' });
+
+const RETRY_COUNT = 30;
+const POLL_INTERVAL = 2000;
+
+class DockerService {
+  /**
+   * Prepares the docker testing environment.
+   */
+  async onPrepare(config) {
+    this.host = config.hostname;
+    this.port = config.port;
+
+    await this.initializeSwarm();
+    await this.deployStack();
+  }
+
+  /**
+   * Initializes a docker swarm instance.
+   */
+  async initializeSwarm() {
+    const { stdout: dockerInfo } = await exec('docker info --format "{{json .}}"');
+    const { Swarm } = JSON.parse(dockerInfo);
+
+    if (Swarm.LocalNodeState !== 'active') {
+      logger.log('Initializing docker swarm.');
+
+      await exec('docker swarm init');
+    }
+  }
+
+  /**
+   * Deploys the docker stack.
+   */
+  async deployStack() {
+    // Remove the previous stack if one exists.
+    await this.removeStack();
+
+    logger.log('Deploying docker stack.');
+
+    const composeFilePath = path.resolve(__dirname, '../docker/docker-compose.yml');
+
+    await exec(`docker stack deploy -c ${composeFilePath} wdio`);
+
+    await this.awaitNetworkReady();
+  }
+
+  /**
+   * Removes the docker stack.
+   */
+  async removeStack() {
+    const { stdout: stackInfo } = await exec('docker stack ls | grep wdio || true');
+
+    if (stackInfo) {
+      logger.log('Removing docker stack.');
+
+      await exec('docker stack rm wdio');
```

Comment:
If we wait for this command to complete, why do we need `await this.awaitServiceRemoval();`?

Reply:
`exec` returns a promise. The command is run asynchronously. The promise will resolve when the command has been executed, but the docker network isn't guaranteed to be taken down yet. So this waits for the command to run and then we need to wait for the network/service to be taken down

Reply:
When the service is remove, won't it always be removed, even if the network is still up? I do `this.awaitNetworkRemoval` is needed.

Reply:
If the network is still up when we try to deploy another stack it will cause errors. We have to wait for the network to go down before we create a new one. This happens when the test run is hard-stopped or errors and the onComplete isn't run to take down the stack before launching a new stack. The remove stack will be executed on deploy stack if an existing network is active.

Reply:
ah the hard stopping. Makes sense 👍